docs/source/notes/extending.rst

```diff
@@ -14,59 +14,104 @@ Extending :mod:`torch.autograd`
-  the autograd engine.
-
-  - :meth:`~torch.autograd.function._ContextMethodMixin.save_for_backward` must be
-    used when saving input or output of the forward to be used later in the backward.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.mark_dirty` must be used to
-    mark any input that is modified inplace by the forward function.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.mark_non_differentiable` must
-    be used to tell the engine if an output is not differentiable.
-  - :meth:`~torch.autograd.function._ContextMethodMixin.set_materialize_grads` can be
-    used to tell the autograd engine to optimize gradient computations in the cases where
-    the output does not depend on the input by not materializing grad tensors given to backward
-    function. That is, if set to False, None object in python or "undefined tensor" (tensor x for
-    which x.defined() is False) in C++ will not be converted to a tensor filled with zeros prior
-    to calling backward. However, supporting this optimization means your custom autograd function
-    has to handle gradients that are represented in this way and is thus opt-in. Default value is True.
-
-.. note::
-
-  By default, all the output Tensors that are of differentiable type will be set to
-  require gradient and have all autograd metadata set for them. If you don't want
-  them to require gradients, you can use the `mark_non_differentiable` method mentioned
-  above. For output Tensors that are not of differentiable type (integer types for example),
-  they won't be marked as requiring gradients.
+**Step 2:** It is your responsibility to use the functions in the forward's `ctx`
+properly in order to ensure that the new :class:`Function` works properly with
+the autograd engine.
+
+- :meth:`~torch.autograd.function.FunctionCtx.save_for_backward` must be
+  used when saving input or output of the forward to be used later in the backward.
+- :meth:`~torch.autograd.function.FunctionCtx.mark_dirty` must be used to
+  mark any input that is modified inplace by the forward function.
+- :meth:`~torch.autograd.function.FunctionCtx.mark_non_differentiable` must
+  be used to tell the engine if an output is not differentiable. By
+  default all output tensors that are of differentiable type will be set
+  to require gradient. Tensors of non-differentiable type (i.e., integral types)
+  are always marked as requiring gradients.
+- :meth:`~torch.autograd.function.FunctionCtx.set_materialize_grads` can be
+  used to tell the autograd engine to optimize gradient computations in the cases where
+  the output does not depend on the input by not materializing grad tensors given to backward
+  function. That is, if set to False, None object in python or "undefined tensor" (tensor x for
+  which x.defined() is False) in C++ will not be converted to a tensor filled with zeros prior
+  to calling backward. However, supporting this optimization means your custom autograd function
```

Comment:
nit I would add at the end: And so your code will need to handle such objects as if they were Tensors filled with zeros.

Reply:
It slightly repeats the sentence: "supporting this optimization means your custom autograd function has to handle gradients that are represented in this way", so I just removed that one.